mount/mount_linux.go

```diff
@@ -324,10 +352,27 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
 		}
 
 		klog.Infof("Disk successfully formatted (mkfs): %s - %s %s", fstype, source, target)
-	} else if fstype != existingFormat {
-		// Verify that the disk is formatted with filesystem type we are expecting
-		mountErrorValue = FilesystemMismatch
-		klog.Warningf("Configured to mount disk %s as %s but current format is %s, things might break", source, existingFormat, fstype)
+	} else {
+		if fstype != existingFormat {
+			// Verify that the disk is formatted with filesystem type we are expecting
+			mountErrorValue = FilesystemMismatch
+			klog.Warningf("Configured to mount disk %s as %s but current format is %s, things might break", source, existingFormat, fstype)
+		}
+
+		if !readOnly {
+			// Run check tools on the disk to fix repairable issues, only do this for formatted volumes requested as rw.
+			var err error
+			switch existingFormat {
+			case "xfs":
+				err = mounter.checkAndRepairXfsFilesystem(source)
+			default:
+				err = mounter.checkAndRepairFilesystem(source)
+			}
+
+			if err != nil {
+				return err
```

Comment:
Rebase is not that simple in this case: you should return `NewMountError(mountErrorValue, err.Error())`, to propagate potential `FilesystemMismatch`.

Reply:
Actually, you should return `FilesystemMismatch`, if the FS type is wrong, or `HasFilesystemErrors` if the FS type is OK, but the filesystem has errors.

Reply:
1. HasFilesystemErrors is retruned in checkAndRepairXfsFilesystem/checkAndRepairFilesystem, so it is ok to return err here directly.
2. I don't think we should return a FilesystemMismatch here, because the only error which will happen here is a failed_to_repair_filesystem_error, and the fs passed in is the existing fs, not the declear one, so I think there is no filesystem mismatch which can cause this error.
3. I forgot to return a HasFilesystemErrors in checkAndRepairXfsFilesystem, which is fixed now.

Reply:
ack

Reply:
ack, I missed checkAndRepairXfsFilesystem/checkAndRepairFilesystem